### PR TITLE
[ECO-643] Implement Observables Handler

### PR DESCRIPTION
**Pull request by imonaco**
### What ?

Implement methods to interact with observables related operations.

### Why ?

We are lacking a handler to interact with the observables API

### How ?

Creating a new handler, and refactor a bit. Also add more test cases!

### CODE REVIEW

File: tests/unit/test_submission.py [hunk 1]
```diff
@@ -342,3 +283,9 @@ def test_get_submission_status(submission, mocked_request):
     mocked_request.get(expected_url, json=json_response)
     response = submission.get_submission_status("test-submission-id")
     assert response.json() == json_response
+
+def test_set_sort_order_not_implemented(submission):
+    assert len(submission.payload_params) == TOTAL_DEFAULT_PARAMS
+    with pytest.raises(NotImplementedError):
+        submission.set_sort_order("ASC")
+    assert len(submission.payload_params) == TOTAL_DEFAULT_PARAMS
```
Comment (hernantain) [inline tests/unit/test_submission.py:291]: EOF

File: trustar2/handlers/search_handler.py [hunk 2]
```diff
@@ -46,42 +45,49 @@ def _validate_dates(self):
...
         self.set_payload_param("sortColumn", column)
+
+
+    def set_sort_order(self, order):
+        order = self._get_value(order, SortOrder)
+        self.set_payload_param("sortOrder", order)
+
+
+
+
```
Comment (hernantain) [inline trustar2/handlers/search_handler.py:93]: Empty lines

File: trustar2/handlers/submission.py [hunk 3]
```diff
@@ -258,3 +258,6 @@ def get_submission_status(self, submission_id):
     def tags(self):
         """Returns a TagSubmission handler that will let you alter-tags on a submission"""
         return TagSubmission(self.config)
+
+    def set_sort_order(self, order):
+        raise NotImplementedError("Submissions dont support sort order")
```
Comment (hernantain) [inline trustar2/handlers/submission.py:263]: EOF

File: trustar2/handlers/tags.py [hunk 4]
```diff
@@ -94,3 +95,19 @@ def set_submission_id(self, submission_id):
...
+
+
+    def set_observable_value(self, value):
+        self.guid = value
+        self.set_payload_param("observableValue", value)
+
+    @property
+    def tag_endpoint(self):
+        self._validate_payload()
+        return "{}{}".format(self.base_url, "/alter-tags")
```
Comment (hernantain) [inline trustar2/handlers/tags.py:113]: eof

File: trustar2/trustar_enums.py [hunk 5]
```diff
@@ -66,3 +66,8 @@ class WorkflowDestinations(TSEnum):
     QRADAR = "QRADAR"
     ENCLAVE = "ENCLAVE"
     UNKNOWN = "UNKNOWN"
+
+class ID_Types(TSEnum):
+    INTERNAL = "INTERNAL"
+    EXTERNAL = "EXTERNAL"
+    UNRECOGNIZED = "UNRECOGNIZED"
```
Comment (hernantain) [inline trustar2/trustar_enums.py:73]: eof

File: trustar2/handlers/observables.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    def set_sort_column(self, column):
+        """
+        Specify the column to sort the search results.
+
+        :param column: 'FIRST_SEEN' | 'LAST_SEEN'
+        :returns: self.
+        """
+        class ObservableSortColumns(TSEnum):
+            FIRST_SEEN = "FIRST_SEEN"
+            LAST_SEEN = "LAST_SEEN"
```
Comment (hernantain) [inline trustar2/handlers/observables.py:68]: Can we move this one to: `trustar2/trustar_enums.py`

File: trustar2/handlers/observables.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        return query.set_query_string(self.query_params)
+
+    def set_search_types(self, types):
+        """
+        Adds types to the search filters.
+
+        :param types: Array of ObesrvableTypes Values
+        :returns: self.
+        """
+        types = list(map(lambda t: self._get_value(t, ObservableTypes), set(types)))
```
Comment (hernantain) [inline trustar2/handlers/observables.py:56]: nice

File: trustar2/handlers/observables.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    def tags(self):
+        """
+        Returns an observable tags handler
+
+        :returns: TagObservable handler.
+        """
+        return TagObservable(self.config)
+
+    def _validate_tags_length(self):
+        MAX_TAG_LENGTH = 20
```
Comment (hernantain) [inline trustar2/handlers/observables.py:80]: We have an enum with this value [here](https://github.com/trustar/trustar-sdk2-proto/blob/main/trustar2/trustar_enums.py#L58)

File: trustar2/handlers/search_handler.py [hunk 9]
```diff
@@ -46,42 +45,49 @@ def _validate_dates(self):
...
 
+    def _argument_to_list(self, arg):
+        iterables = (list, tuple, set)
+        return arg if isinstance(arg, iterables) else [arg]
+    
+    def _argument_to_unique_list(self, arg):
+        return list(set(self._argument_to_list(arg)))
+
+    def _process_date(self, date):
+        return date if isinstance(date, int) else get_timestamp(date)
```
Comment (hernantain) [inline trustar2/handlers/search_handler.py:56]: Nice! Can we move this methods into the `BaseHandler` class? This is applicable for lots of things not only search params.

File: tests/unit/test_tags.py [hunk 10]
```diff
@@ -10,22 +11,23 @@
 
 ENCLAVE_ID = "3a93fab3-f87a-407a-9376-8eb3fae99b4e"
 IOC_SUBMISSION_GUID = "cc12a5c6-e575-3879-8e41-2bf240cc6fce"
-EXPECTED_URL = BASE_URL.format("/{}/cc12a5c6-e575-3879-8e41-2bf240cc6fce/alter-tags")
-
+OBSERVABLE_VALUE = "b1fa0ef4930abc0c681081ef2d2f834b3b2fbbbd"
+BASE_URL = "https://api.trustar.co/api/2.0/"
+EXPECTED_URL = BASE_URL + "{}" + "/{}/alter-tags".format(IOC_SUBMISSION_GUID)
```
Comment (hernantain) [inline tests/unit/test_tags.py:16]: I did a last minute change of this in my last PR and included a BASE_URL string in `conftest.py`

File: tests/conftest.py [hunk 11]
```diff
@@ -12,3 +12,11 @@ def mocked_request():
...
         m.post(url="/oauth/token", text='{"access_token": "TOKEN12345"}')
         yield m
+
+@pytest.fixture
+def ts():
+    return TruStar(
+        api_key="xxxx",
+        api_secret="xxx",
+        client_metatag="test_env"
+    )
```
Comment (hernantain) [inline tests/conftest.py:22]: Nice

File: tests/unit/resources.py [hunk 12]
```diff
@@ -227,3 +227,13 @@
...
+
+observables_search_example_request = """{
+    "queryTerm": "query",
+    "from":1596607968000,
+    "to":1598308171000,
+    "sortColumn": "FIRST_SEEN",
+    "sortOrder": "ASC",
+    "enclaveGuids": ["4bdc3f5b-3ed5-4d99-b20c-2d801866ef0b"],
+    "types": ["MD5"]
+}"""
```
Comment (hernantain) [inline tests/unit/resources.py:239]: eof

File: tests/unit/test_observables_handler.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        .set_sort_column("FIRST_SEEN")
+        .set_sort_order("ASC")
+        .search()
+    )
+    assert q.params.serialize() == json.loads(observables_search_example_request)
+
+def test_get_from_submission(observables_handler):
+    submission_id = "2d87abca-3b6c-4bdc-b9dd-0a3e29901bbb"
+    q = observables_handler.get_from_submission(submission_id)
+    assert(q.query_string.get("submissionId") ==  submission_id)
```
Comment (hernantain) [inline tests/unit/test_observables_handler.py:91]: EOF

File: tests/unit/test_observables_handler.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+def test_search_too_many_included_tags(observables_handler):
+    MAX_TAGS = 20
+    tags = list(range(0, MAX_TAGS + 1))
+    observables_handler.set_included_tags(tags)
+    with pytest.raises(AttributeError):
+        observables_handler.search()
+
+def test_search_too_many_excluded_tags(observables_handler):
+    MAX_TAGS = 20
```
Comment (hernantain) [inline tests/unit/test_observables_handler.py:56]: Can we move MAX_TAGS to a constant at module level so we don't have to repeat it everytime? We can even do something like:

```python
from trustar2.trustar_enums import MaxValues

MAX_TAGS = MaxValues.TAGS.value
```

File: tests/unit/test_tags.py [hunk 15]
```diff
@@ -125,3 +128,70 @@ def test_alter_tag_submissions_incomplete_with_missing_enclave_id(tag_submission
...
+    ("added_tags, removed_tags"),
+    (
+        (["important", "tag"], []),
+        ([], ["not_important", "tag"]),
+        (["important", "tag"], ["not_important"])
+    )
+)
+def test_ok_alter_tags_observables(tag_observable, mocked_request, added_tags, removed_tags):
+
+    EXPECTED_URL = "{}{}/{}".format(BASE_URL, "observables", "alter-tags")
```
Comment (hernantain) [inline tests/unit/test_tags.py:145]: Should be also changed here.

File: trustar2/handlers/search_handler.py (hunk 9 above)
Comment (hernantain) [reply trustar2/handlers/search_handler.py:56]: Also, maybe you can move the logic of `_process_date` to [here](https://github.com/trustar/trustar-sdk2-proto/blob/main/trustar2/base.py#L12)

File: trustar2/handlers/tags.py [hunk 16]
```diff
@@ -94,3 +95,19 @@ def set_submission_id(self, submission_id):
 
     def set_id_type_as_external(self, external):
         self.set_payload_param("idType", "EXTERNAL" if external else "INTERNAL")
+
+@fluent
+class TagObservable(TagBase):
+
```
Comment (hernantain) [inline trustar2/handlers/tags.py:101]: Can you add this import on `trustar2.__init__.py` and `trustar2.handlers.__init__.py` ?

**Review by hernantain — CHANGES_REQUESTED** [review]
Great Work @imonaco 💯 ! Looking good! Just have some minor comments. (Sorry for all the newline at "eof" comments, just want github to stop complaining)

File: tests/unit/test_submission.py (hunk 1 above)
Comment (imonaco) [reply tests/unit/test_submission.py:291]: yeah, mb haha, fixing all them in a minute

Comment (imonaco) [reply tests/unit/test_submission.py:291]: we should consider implementing a linter such as black

File: tests/unit/test_observables_handler.py (hunk 14 above)
Comment (imonaco) [reply tests/unit/test_observables_handler.py:56]: Yeah, i wanted to do that, but then i thought that not all the tags for every entity has 20 as max, any suggestions about this issue? or all has the same?

File: trustar2/handlers/tags.py (hunk 16 above)
Comment (imonaco) [reply trustar2/handlers/tags.py:101]: Oh right, the imports, sorry, total forgot about that

File: tests/unit/test_observables_handler.py (hunk 14 above)
Comment (imonaco) [reply tests/unit/test_observables_handler.py:56]: I just look into the enum file and it already exists, lol, im changing it right now

File: tests/conftest.py (hunk 11 above)
Comment (marcos-vivar) [reply tests/conftest.py:22]: Nice indeed!

**Review by marcos-vivar — APPROVED** [review]
Outstanding work @imonaco! 👍🏽 

Minor comments but LGTM! :shipit:

File: trustar2/handlers/__init__.py [hunk 17]
```diff
@@ -1,6 +1,7 @@
-from trustar2.handlers.tags import TagIndicator, TagSubmission
+from trustar2.handlers.tags import TagIndicator, TagSubmission, TagObservable
 from trustar2.handlers.indicators import SearchIndicator
 from trustar2.handlers.safelist import Safelist
 from trustar2.handlers.submission import Submission
 from trustar2.handlers.account import Account
 from trustar2.handlers.workflow import Workflow
+from trustar2.handlers.observables import ObservablesHandler
```
[the remark this answers is not in the document]
Comment (imonaco) [reply trustar2/handlers/__init__.py:7]: omg , they keep appearing haha

File: trustar2/handlers/observables.py (hunk 6 above)
Comment (imonaco) [reply trustar2/handlers/observables.py:70]: Yes!

File: trustar2/handlers/search_handler.py (hunk 9 above)
Comment (imonaco) [reply trustar2/handlers/search_handler.py:56]: Will do

File: tests/unit/test_tags.py (hunk 10 above)
Comment (imonaco) [reply tests/unit/test_tags.py:16]: let me check this

Comment (imonaco) [reply tests/unit/test_tags.py]: done

File: trustar2/handlers/observables.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+from trustar2.query import Query
+from trustar2.base import fluent, Methods
+from trustar2.handlers.tags import TagObservable
+from trustar2.handlers.search_handler import SearchHandler
+from trustar2.trustar_enums import ID_Types, ObservableSortColumns, ObservableTypes, MaxValues
+
+MAX_TAGS = MaxValues.TAGS.value 
+
+@fluent
+class ObservablesHandler(SearchHandler):
```
Comment (hernantain) [inline trustar2/handlers/observables.py:12]: Just realized, shouldn't we be naming this just `Observables` to have consistency with all handlers?

Comment (imonaco) [reply trustar2/handlers/observables.py:12]: Yeah... I wanted to have a talk with you about naming convention, it seems to be a bit irregular, but yes, you are right.

**Review by hernantain — APPROVED** [review]
Great work @imonaco ! LGTM

**Pull request merged**